src/com/ms/silverking/cloud/zookeeper/ZooKeeperExtended.java

```diff
@@ -42,1024 +43,1081 @@
+        } else {
+          if (retrySW.getSplitSeconds() > reissueIntervalSeconds) {
+            if (timeoutSW.getSplitSeconds() > timeoutSeconds) {
+              markIncompleteAsErrors();
+            } else {
+              issueGetsForIncomplete();
             }
+          }
         }
-    }
-    
-    class Result {
-        private final AsyncGet    asyncGet;
-        private final int        rc;
-        private final String    path;
-        private final Object    ctx;
-        private final byte[]    data;
-        private final Stat        stat;
-        
-        Result(AsyncGet asyncGet, int rc, String path, Object ctx, byte[] data, Stat stat) {
-            this.asyncGet = asyncGet;
-            this.rc = rc;
-            this.path = path;
-            this.ctx = ctx;
-            this.data = data;
-            this.stat = stat;
-        }
-    }        
-
-    //////////////////////////////////////////
-    
-    private Map<String,String> convertToStringData(Map<String,byte[]> data) {
-        Map<String,String>    stringData;
-        
-        stringData = new HashMap<>();
-        for (Map.Entry<String, byte[]> e : data.entrySet()) {
-            stringData.put(e.getKey(), new String(e.getValue()));
-        }
-        return stringData;
-    }
-    
-    private Map<String, Integer> convertToIntData(Map<String, byte[]> data) {
-        Map<String,Integer>    intData;
-        
-        intData = new HashMap<>();
-        for (Map.Entry<String, byte[]> e : data.entrySet()) {
-            intData.put(e.getKey(), NumConversion.bytesToInt(e.getValue()));
-        }
-        return intData;
-    }
-        
-    public byte[] getByteArray(String path) throws KeeperException {
-        return getByteArray(path, null);
+      }
     }
 
-    public byte[] getByteArray(String path, Watcher watcher) throws KeeperException {
-        return getByteArray(path, watcher, new Stat());
-    }
-    
-    public byte[] getByteArray(String path, Watcher watcher, Stat stat) throws KeeperException {
-        try {
-            byte[]  data;
+    /**
+     * Notify waiter. Let the waiter check for completion, resend etc.
+     * We want to keep callback operations as light as possible.
+     */
+    private void notifyWaiter() {
+      lock.lock();
+      try {
+        cv.signalAll();
+      } finally {
+        lock.unlock();
+      }
 
-            data = getData(path, watcher, stat);
-            return data;
-        } catch (InterruptedException ie) {
-            throw new RuntimeException("panic");
-        }
     }
-    
-    public String getString(String path) throws KeeperException {
-        byte[]  data;
 
-        data = getByteArray(path);
-        return new String(data);
+    @Override
+    public void processResult(int rc, String path, Object ctx, byte[] data, Stat stat) {
+      try {
+        asyncGetResults.put(new Result(this, rc, path, ctx, data, stat));
+      } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)